proxy/http/proxy/server.py

```diff
@@ -175,7 +175,11 @@ async def get_descriptors(self) -> Descriptors:
         return r, w
 
     async def write_to_descriptors(self, w: Writables) -> bool:
-        if (self.upstream and self.upstream.connection.fileno() not in w) or not self.upstream:
+        if (
+            self.upstream
+            and not self.upstream.closed
+            and self.upstream.connection.fileno() not in w
+        ) or not self.upstream:
```

Comment:
Saw following locally. Comparing with `read_from_descriptors` looks like `not self.upstream.closed` was missing within `write_to_descriptors`

```
2024-04-13 13:43:09,964 - pid:71306 [W] server.connect_upstream:607 - Unable to connect with upstream cs.nex8.net:443 due to [Errno 8] nodename nor servname provided, or not known
2024-04-13 13:43:09,964 - pid:71306 [I] handler.handle_data:186 - HttpProtocolException: ProxyConnectionFailed gaierror(8, 'nodename nor servname provided, or not known')
2024-04-13 13:43:09,965 - pid:71306 [E] base_events.default_exception_handler:1785 - Task exception was never retrieved
future: <Task finished name='Task-1' coro=<Threadless._run_forever() done, defined at /Users/abhinavsingh/Dev/proxy.py/proxy/core/work/threadless.py:376> exception=TcpConnectionUninitializedException()>
Traceback (most recent call last):
  File "/Users/abhinavsingh/Dev/proxy.py/proxy/core/work/threadless.py", line 380, in _run_forever
    if await self._run_once():
       ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/abhinavsingh/Dev/proxy.py/proxy/core/work/threadless.py", line 364, in _run_once
    teardown = task.result()
               ^^^^^^^^^^^^^
  File "/Users/abhinavsingh/Dev/proxy.py/proxy/http/handler.py", line 147, in handle_events
    self.writes_teardown = await self.plugin.write_to_descriptors(writables)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/abhinavsingh/Dev/proxy.py/proxy/http/proxy/server.py", line 178, in write_to_descriptors
    if (self.upstream and self.upstream.connection.fileno() not in w) or not self.upstream:
                          ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/abhinavsingh/Dev/proxy.py/proxy/core/connection/server.py", line 32, in connection
    raise TcpConnectionUninitializedException()
proxy.core.connection.connection.TcpConnectionUninitializedException
2024-04-13 13:43:10,374 - pid:71307 [W] server.read_from_descriptors:247 - Connection reset by upstream: gum.criteo.com:443
```